other/upgrade.php

```diff
@@ -715,6 +715,12 @@ function loadEssentialData()
 		return random_bytes($bytes);
 	};
 
+	// This is used in text2words() for old 1.0.x conversions; restoring old logic
+	$smcFunc['truncate'] = function($word, $max_chars)
+	{
+		return substr($word, 0, $max_chars);
+	};
```

Comment:
```suggestion
	$smcFunc['truncate'] = function($word, $max_chars)
	{
		$new_string = '';

		foreach (preg_split('/((?>&.*?;|\X))/u', $string, -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY) as $char)
		{
			if (strlen($new_string . $char) > $max_chars)
				break;

			$new_string .= $char;
		}

		return $new_string;
	};
```


This will avoid chopping in the middle of entities or multi-byte characters. It's still not quite as cushy as the one in Load.php, but it will at least prevent malformed output.